File: src/main/java/io/github/hdzitao/editstarters/buildsystem/gradle/AbstractBuildGradle.java
package io.github.hdzitao.editstarters.buildsystem.gradle;

import com.intellij.psi.PsiElement;
import io.github.hdzitao.editstarters.buildsystem.DependencyScope;
import io.github.hdzitao.editstarters.buildsystem.ProjectFile;
import io.github.hdzitao.editstarters.dependency.Bom;
import io.github.hdzitao.editstarters.dependency.Repository;
import io.github.hdzitao.editstarters.springboot.Starter;
import org.apache.commons.lang3.ArrayUtils;
import org.apache.commons.lang3.StringUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.function.BiFunction;

/**
 * build.gradle抽象类
 *
 * @version 3.2.0
 */
public abstract class AbstractBuildGradle<T extends PsiElement> extends ProjectFile<T> {

    /**
     * gradle语法简单抽象
     */
    public static class Instruction {
        private final String inst;
        private final String point;

        public Instruction(String inst, String point) {
            this.inst = inst;
            this.point = point;
        }

        public String getInst() {
            return inst;
        }

        public String toInstString(String format) {
            return format.replace("${inst}", inst).replace("${point}", point);
        }
    }

    // TAG ==============================================================================
    public static final String TAG_DEPENDENCY_MANAGEMENT = "dependencies";
    public static final String TAG_BOM_MANAGEMENT = "dependencyManagement";
    public static final String TAG_BOM_IMPORT = "imports";
    public static final String TAG_BOM = "mavenBom";
    public static final String TAG_REPOSITORY_MANAGEMENT = "repositories";
    public static final String TAG_REPOSITORY = "maven";
    // TAG ==============================================================================

    /**
     * dependency语法
     */
    protected List<Instruction> dependencyInstruction(Starter info) {
        List<Instruction> instructions = new ArrayList<>();
        String point = splicingDependency(info.getGroupId(), info.getArtifactId(), info.getVersion());
        for (String inst : resolveScope(DependencyScope.getByScope(info.getScope()))) {
            instructions.add(new Instruction(inst, point));
        }
        return instructions;
    }

    /**
     * bom语法
     */
    protected Instruction bomInstruction(Bom bom) {
        String point = splicingDependency(bom.getGroupId(), bom.getArtifactId(), bom.getVersion());
        return new Instruction(TAG_BOM, point);
    }

    /**
     * repository语句
     */
    protected Instruction repositoryInstruction(Repository repository) {
        return new Instruction(TAG_REPOSITORY, repository.getUrl());
    }

    /**
     * 割出GroupID/ArtifactID构建Depend
     */
    protected <Depend> Depend newByGroupArtifact(String point, BiFunction<String, String, Depend> buildFun) {
        if (StringUtils.isNoneBlank(point)) {
            String[] groupArtifact = point.split(":");
            if (groupArtifact.length >= 2) {
                return buildFun.apply(groupArtifact[0], groupArtifact[1]);
            }
        }
        return buildFun.apply(EMPTY, EMPTY);
    }

    /**
     * 处理scope
     */
    protected String[] resolveScope(DependencyScope scope) {
        return switch (scope) {
            case ANNOTATION_PROCESSOR -> new String[]{"compileOnly", "annotationProcessor"};
            case COMPILE -> new String[]{"implementation"};
            case COMPILE_ONLY -> new String[]{"compileOnly"};
            case PROVIDED -> new String[]{"providedRuntime"};
            case RUNTIME -> new String[]{"runtimeOnly"};
            case TEST -> new String[]{"testImplementation"};
        };
    }

    /**
     * " aaa " => "aaa"
     */
    protected String trimText(String s, char... chars) {
        if (s == null) {
            return EMPTY;
        }

        int len = s.length();
        int st = 0;
        char[] val = s.toCharArray();

        while ((st < len) && (ArrayUtils.contains(chars, val[st]))) {
            st++;
        }
        while ((st < len) && (ArrayUtils.contains(chars, val[len - 1]))) {
            len--;
        }
        return ((st > 0) || (len < s.length())) ? s.substring(st, len) : s;
    }

    /**
     * 检查空
     */
    protected String checkEmpty(String s) {
        return s != null ? s : EMPTY;
    }

    /**
     * 拼接 groupId:artifactId:version
     */
    protected String splicingDependency(String groupId, String artifactId, String version) {
        groupId = checkEmpty(groupId);
        artifactId = checkEmpty(artifactId);

        String point = groupId + ":" + artifactId;

        if (StringUtils.isNoneBlank(version)) {
            point = point + ":" + version;
        }

        return point;
    }

}

File: src/main/kotlin/io/github/hdzitao/editstarters/buildsystem/gradle/AbstractBuildGradle.kt
package io.github.hdzitao.editstarters.buildsystem.gradle

import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import io.github.hdzitao.editstarters.buildsystem.DependencyScope
import io.github.hdzitao.editstarters.buildsystem.ProjectFile
import io.github.hdzitao.editstarters.dependency.Bom
import io.github.hdzitao.editstarters.dependency.Repository
import io.github.hdzitao.editstarters.springboot.Starter
import org.apache.commons.lang3.ArrayUtils

/**
 * build.gradle抽象类
 *
 * @version 3.2.0
 */
abstract class AbstractBuildGradle<BuildFile : PsiFile, Psi : PsiElement> : ProjectFile<BuildFile, Psi>() {
    /**
     * gradle语法简单抽象
     */
    class Instruction(val inst: String, val point: String) {
        fun toInstString(format: String): String {
            return format.replace("\${inst}", inst).replace("\${point}", point)
        }
    }

    companion object {
        // TAG ==============================================================================
        const val TAG_DEPENDENCY_MANAGEMENT: String = "dependencies"
        const val TAG_BOM_MANAGEMENT: String = "dependencyManagement"
        const val TAG_BOM_IMPORT: String = "imports"
        const val TAG_BOM: String = "mavenBom"
        const val TAG_REPOSITORY_MANAGEMENT: String = "repositories"
        const val TAG_REPOSITORY: String = "maven"
        // TAG ==============================================================================
    }

    /**
     * dependency语法
     */
    protected fun dependencyInstruction(info: Starter): List<Instruction> {
        val instructions = ArrayList<Instruction>()
        val point = splicingPoint(info.groupId, info.artifactId, info.version)
        for (inst in DependencyScope.getByScope(info.scope).resolveGradleScope()) {
            instructions.add(Instruction(inst, point))
        }
        return instructions
    }

    /**
     * bom语法
     */
    protected fun bomInstruction(bom: Bom): Instruction {
        val point = splicingPoint(bom.groupId, bom.artifactId, bom.version)
        return Instruction(TAG_BOM, point)
    }

    /**
     * repository语句
     */
    protected fun repositoryInstruction(repository: Repository): Instruction {
        val point = repository.url ?: EMPTY
        return Instruction(TAG_REPOSITORY, point)
    }

    /**
     * 割出GroupID/ArtifactID构建Depend
     */
    protected fun <Depend> newByGroupArtifact(point: String, buildFun: (String, String) -> Depend): Depend {
        if (point.isNotBlank()) {
            val groupArtifact = point.split(":")
            if (groupArtifact.size >= 2) {
                return buildFun(groupArtifact[0], groupArtifact[1])
            }
        }
        return buildFun(EMPTY, EMPTY)
    }

    /**
     * 处理scope
     */
    protected fun DependencyScope.resolveGradleScope(): Array<String> {
        return when (this) {
            DependencyScope.ANNOTATION_PROCESSOR -> arrayOf("compileOnly", "annotationProcessor")
            DependencyScope.COMPILE -> arrayOf("implementation")
            DependencyScope.COMPILE_ONLY -> arrayOf("compileOnly")
            DependencyScope.PROVIDED -> arrayOf("providedRuntime")
            DependencyScope.RUNTIME -> arrayOf("runtimeOnly")
            DependencyScope.TEST -> arrayOf("testImplementation")
        }
    }

    /**
     * " aaa " => "aaa"
     */
    protected fun String?.trimText(vararg chars: Char): String {
        if (this == null) {
            return EMPTY
        }

        var len = length
        var st = 0
        val v = toCharArray()

        while ((st < len) && (ArrayUtils.contains(chars, v[st]))) {
            st++
        }
        while ((st < len) && (ArrayUtils.contains(chars, v[len - 1]))) {
            len--
        }
        return if (st > 0 || len < length) substring(st, len) else this
    }

    /**
     * 检查空
     */
    protected fun String?.checkEmpty(): String {
        return this ?: EMPTY
    }

    /**
     * 拼接 groupId:artifactId:version
     */
    protected fun splicingPoint(groupId: String?, artifactId: String?, version: String?): String {
        return "${groupId.checkEmpty()}:${artifactId.checkEmpty()}${version?.let { ":$it" } ?: ""}"
    }
}